### PR TITLE
usage of sparsemax in enc-dec attention task

### PR DESCRIPTION
**Pull request by AndreasMadsen**
**rebase of #8**

An ipython notebook for usage test of softmax vs. sparsemax.

In the notebook, search for `attention_fn`, replace the current argument with the `sparse_max` function instead of the `tf.nn.softmax` function.

You might want to start with having the `sparse_max` outputting a dense tensor to test it works as intended.
Having the output as a [SparseTensor](https://www.tensorflow.org/versions/r0.11/api_docs/python/sparse_ops.html#sparse-tensors) might cause some issues.

### CODE REVIEW

**Comment by FrederikWR** [comment]
The Ipython notebook now contains the graphs/output of actually running the code. As can be seen in the plots the sparse model has better performance as well as actual sparse attention (avg. number of attention elements != 33 for sparsemax vs 130 for softmax per row of labels). This can also be seen in the last two histograms.

Note that data_generator has gotten an updated parameter seed in order to run the soft/sparsemax on the same data (I couldn't figure out a workaround to properly reset/initialize/create tensorflow variables on the RNN-class object so the data has to be generated twice). Although it's a bit clunky the notebook works using "reset runall" or by remembering to execute the right reset_default_graph cells.

**Comment by AndreasMadsen** [comment]
> explanation: the tensorflow sparse softmax cross was previously errornously used for both sparse and softmax. after using the appropriate loss function I've yet to observe nan problems - but perhaps i've just been lucky

wow, can you elaborate on what the issue was?

**Comment by FrederikWR** [comment]
Hmm, as I said I'm not completely convinced that the problem is entirely removed. I'm not 100% confident what I did was correct but never the less for sequence of length 5 it seems to work consistenly with the change. Let me know what you think / perhaps we can discuss it in person tomorrow?:

 What I did was that I changed the method "sequence_loss_tensor" in "usage/enc-dec/tf_utils.py". Previously it would use "nn_ops.sparse_softmax_cross_entropy_with_logits" no matter what activation function we were using (i.e. sparse/soft-max). I switched this such that it in the case of a sparsemax activation function also uses the appropriate sparsemax loss. Amongst other things this requires recast the vector of labels into the 1-hot encoding expected by our sparsemax loss function

In addition, some googling shows that we are not the first people to experience nan using "sparse_softmax_cross_entropy_with_logits"

**Comment by AndreasMadsen** [comment]
Ah yes, `log(0)` will return `nan`, maybe we are getting `0` when using sparsemax. A solution is sometimes to add epsilon to the loss input.
